### PR TITLE
Ruby 3.0 Support

### PR DESCRIPTION
**Pull request by alexhemard**
This PR:

1. Makes `smash-the-state` compatible with Ruby 3.1 
2. Contains a handful of `rubocop` updates
3. Removes the freeze on original state

The freeze is removed because it is not compatible with operations that contain an `attr_accessor` in the schema. For example:

```
 ...

schema do
  schema :deployment do
    attribute :type, :string
  end

  attr_accessor :private_endpoints_available
end

step :add_private_endpoints_available do |endpoints, params|
  deployment = ...
  params.private_endpoints_available = deployment.private_endpoints_available
  
  endpoints
end
```

Results in:

```
Failure/Error: params.private_endpoints_available = deployment.private_endpoints_available

FrozenError: can't modify frozen #<Class:0x0000000110fb6808>:
...
```

@onyxrev you mentioned briefly that these incompatible operations could be rewritten to make them work, but I'm not sure how that would work. The example above is extracted from a real operation that defines the schema, performs validation, and then attempts to assign the accessor in the second step. The accessor is attached to the original state, which gets passed to each step, which allows us to assign the accessor at any time in the operation. But when the original state is frozen, the accessor becomes unusable.

My preference would be to remove frozen as proposed for now so we aren't blocked, and if in the future we find a way to freeze the original state without losing the accessors, revisit.

### CODE REVIEW

**Comment by alexhemard** [comment]
If the goal is to preserve the original state, would passing a duplicate to each step might be an acceptable compromise? i.e. changing this 
https://github.com/IBM-Cloud/smash-the-state/blob/6106fb540ed28e76aaf90fdb0a904f9c1c242c49/lib/smash_the_state/operation/sequence.rb#L156

to something like this
```ruby
# pass duplicate of original state for each step
 s.implementation.call(memo, make_original_state(state), run_options)
```

Or some option to operation to allow mutating the original state.

**Comment by onyxrev** [comment]
> If the goal is to preserve the original state, would passing a duplicate to each step might be an acceptable compromise? i.e. changing this
> 
> https://github.com/IBM-Cloud/smash-the-state/blob/6106fb540ed28e76aaf90fdb0a904f9c1c242c49/lib/smash_the_state/operation/sequence.rb#L156
> 
> to something like this
> 
> ```ruby
> # pass duplicate of original state for each step
>  s.implementation.call(memo, make_original_state(state), run_options)
> ```
> 
> Or some option to operation to allow mutating the original state.

I believe it is dup'd already? If I remember correctly and am reading the code right, original_state *is* `make_original_state(state)`. It's assigned further up and is defined via the block scope closure.

https://github.com/IBM-Cloud/smash-the-state/blob/6106fb540ed28e76aaf90fdb0a904f9c1c242c49/lib/smash_the_state/operation/sequence.rb#L137

Sure, we can keep it dup'd .

**Comment by onyxrev** [comment]
@alexhemard Would you please also bump `active_model_attributes` to `1.6.0` while you're at it and check to make sure IBM's specs still pass?

**Comment by bpeicher** [comment]
@onyxrev  -- @alexhemard says this is good to go on our end and IBM specs are passing. Mind reviewing?

File: smash_the_state.gemspec [hunk 1]
```diff
@@ -10,13 +10,14 @@ Gem::Specification.new do |s|
   s.summary     = "A useful utility for transforming state that provides step " \
                   "sequencing, middleware, and validation."
   s.description = ""
-  s.license     = "Apache-2.0"
+  s.required_ruby_version = '>= 3.1'
```
Comment (onyxrev) [inline smash_the_state.gemspec]: Is it necessary to require ruby 3?

**Review by onyxrev — APPROVED** [review]
Looks good! Thanks for this :) One question about ruby 3 requirement.

File: smash_the_state.gemspec (hunk 1 above)
Comment (bpeicher) [reply smash_the_state.gemspec]: @onyxrev not necessary. Removed ✔️

**Pull request merged**